File: Data Structures & Algorithms/redundant-connection/submission-1.py
class UnionFind:
    def __init__(self, n):
        self.parents = [i for i in range(n+1)]
        self.ranks = [0] * (n+1)
    
    def find(self,p:int):
        if p != self.parents[p]:
            self.parents[p] = self.find(self.parents[p])
        
        return self.parents[p]
    
    def union(self,p1:int, p2:int) -> bool:
        r1 = self.find(p1)
        r2 = self.find(p2)

        if r1 == r2:
            return False
        
        if self.ranks[r1] >= self.ranks[r2]:
            self.parents[r2] = r1
            self.ranks[r1] += 1

        else:
            self.parents[r1] = r2
            self.ranks[r2] += 1
        
        return True


class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        uf = UnionFind(len(edges))
        for from_node, to_node in edges:
            if not uf.union(from_node,to_node):
                return [from_node,to_node]
        
        return []

File: Data Structures & Algorithms/count-connected-components/submission-1.cpp
class UnionFind {
    public:
        int n;
        vector<int>parents;
        UnionFind(int _n): n(_n) {
            parents.resize(n,0);
            for(int i=0; i<n; i++){
                parents[i] = i; //each node is a parent of itself
            }
        };

        int Find(int v){
            if(parents[v] == v){
                return v;
            }
            return Find(parents[v]);
        }

        void Union(int parent,int child){
            int a = Find(parent);
            int b = Find(child);

            parents[b] = parents[a];
        }

        int getDistinctParents(){
            return set<int>(parents.begin(),parents.end()).size();
        }
};

class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        /*
            0 - 1 - 2 - 3 - 4- 5

            0 1 2 3 4 5
        */
        UnionFind *uf = new UnionFind(n);
        for(vector<int>edge : edges){
            uf->Union(edge[0],edge[1]);
        }
        set<int>s;
        for(auto i: uf->parents){
            s.insert(uf->Find(i));
        }
        cout<<endl;
        return s.size();
    }
};
